src/Tools/DependencyManagement/AbstractServiceProvider.php

```diff
@@ -0,0 +1,109 @@
+	 * @param bool   $shared Whether to register the class as shared (`get` always returns the same instance) or not.
+	 *
+	 * @return DefinitionInterface The generated container definition.
+	 *
+	 * @throws \Exception Error when reflecting the class, or class constructor is not public, or an argument has no valid type hint.
+	 */
+	public function addWithAutoArguments( string $class_name, bool $shared = false ) : DefinitionInterface {
+		try {
+			$reflector = new \ReflectionClass( $class_name );
+		} catch ( \ReflectionException $ex ) {
+			throw new \Exception( get_class( $this ) . "::addWithAutoArguments: error when reflecting class '$class_name': {$ex->getMessage()}" );
+		}
+
+		$definition = new Definition( $class_name, null );
+
+		$constructor = $reflector->getConstructor();
+
+		if ( ! is_null( $constructor ) ) {
+			if ( ! $constructor->isPublic() ) {
+				throw new \Exception( get_class( $this ) . "::addWithAutoArguments: constructor of class '$class_name' isn't public, instances can't be created." );
+			}
+
+			$constructor_arguments = $constructor->getParameters();
+			foreach ( $constructor_arguments as $argument ) {
+				$argument_class = $argument->getClass();
+				if ( is_null( $argument_class ) ) {
+					throw new \Exception( get_class( $this ) . "::addWithAutoArguments: constructor argument '{$argument->getName()}' of class '$class_name' doesn't have a type hint or has one that doesn't specify a class." );
+				}
+
+				$definition->addArgument( $argument_class->name );
+			}
+		}
+
+		// Register the definition only after being sure that no exception will be thrown.
+
+		$this->getContainer()->add( $definition->getAlias(), $definition, $shared );
+
+		return $definition;
+	}
+
+	/**
+	 * Register a class in the container and use reflection to guess the constructor arguments.
+	 * The class is registered as shared, so `get` on the container always returns the same instance.
+	 *
+	 * @param string $class_name Class name to register.
+	 *
+	 * @return DefinitionInterface The generated container definition.
+	 *
+	 * @throws \Exception Error when reflecting the class, or class constructor is not public, or an argument has no valid type hint.
+	 */
+	public function shareWithAutoArguments( string $class_name ) : DefinitionInterface {
+		return $this->addWithAutoArguments( $class_name, true );
+	}
+
+	/**
+	 * Register an entry in the container.
+	 *
+	 * @param string     $id Entry id (typically a class or interface name).
+	 * @param mixed|null $concrete Concrete entity to register under that id, null for automatic creation.
+	 * @param bool|null  $shared Whether to register the class as shared (`get` always returns the same instance) or not.
+	 *
+	 * @return DefinitionInterface The generated container definition.
+	 */
+	public function add( string $id, $concrete = null, bool $shared = null ) : DefinitionInterface {
+		return $this->getContainer()->add( $id, $concrete, $shared );
```

Comment:
Yes, but how? Disallowing anything not in the `Automattic\WooCommerce` namespace?

Reply:
Since any composer package should have a namespace, we could likely just get away with disallowing anything in the root namespace. There are some PHP and WordPress classes that perhaps might be affected by it but I wonder if we could get away with it using a whitelist too. Do you think this is overboard perhaps and its fine the way it is?

Reply:
It's safer to start being too restrictive and relax rules later if needed (which would be easy to do) than to later discover that we were too permissive (which would be impossible to fix). So let's start by just allowing registering classes in `Automattic\Woocommerce` (except the three namespaces that aren't actually ours)?

Reply:
I think that's a fair place to start.